client/fingerprint/network_unix.go

```diff
@@ -152,99 +161,83 @@ func (f *NetworkFingerprint) linkSpeedEthtool(path, device string) int {
 	return mbs
 }
 
-// ipAddress returns the first IPv4 address on the configured default interface
-// Tries Golang native functions and falls back onto ifconfig
-func (f *NetworkFingerprint) ipAddress(device string) string {
-	if ip, err := f.nativeIpAddress(device); err == nil {
-		return ip
-	}
+// Gets the ipv4 addr for a network interface
+func (f *NetworkFingerprint) ipAddress(intf *net.Interface) (string, error) {
+	var addrs []net.Addr
+	var err error
 
-	return f.ifConfig(device)
-}
-
-func (f *NetworkFingerprint) nativeIpAddress(device string) (string, error) {
-	// Find IP address on configured interface
-	var ip string
-	ifaces, err := net.Interfaces()
-	if err != nil {
-		return "", errors.New("could not retrieve interface list")
+	if addrs, err = f.interfaceDetector.Addrs(intf); err != nil {
+		return "", err
 	}
 
-	// TODO: should we handle IPv6 here? How do we determine precedence?
-	for _, i := range ifaces {
-		if i.Name != device {
-			continue
-		}
-
-		addrs, err := i.Addrs()
-		if err != nil {
-			return "", errors.New("could not retrieve interface IP addresses")
+	if len(addrs) == 0 {
+		return "", errors.New(fmt.Sprintf("Interface %s has no IP address", intf.Name))
+	}
+	var ipV4 net.IP
+	for _, addr := range addrs {
+		var ip net.IP
+		switch v := (addr).(type) {
+		case *net.IPNet:
+			ip = v.IP
+		case *net.IPAddr:
+			ip = v.IP
 		}
-
-		for _, a := range addrs {
-			switch v := a.(type) {
-			case *net.IPNet:
-				if v.IP.To4() != nil {
-					ip = v.IP.String()
-				}
-			case *net.IPAddr:
-				if v.IP.To4() != nil {
-					ip = v.IP.String()
-				}
-			}
+		if ip.To4() != nil {
+			ipV4 = ip
+			break
```

Comment:
Instead of break put `return ipV4.String(), nil` here and then after the loop can just be the return err